services/certmanager/standard/service.go

```diff
@@ -0,0 +1,168 @@
+// Copyright © 2020 Attestant Limited.
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package standard
+
+import (
+	"context"
+	"crypto/tls"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/rs/zerolog"
+	zerologger "github.com/rs/zerolog/log"
+	"github.com/wealdtech/go-majordomo"
+)
+
+type Service struct {
+	ctx             context.Context
+	majordomo       majordomo.Service
+	reloadThreshold time.Duration
+	reloadInterval  time.Duration
+	certPEMURI      string
+	certKeyURI      string
+
+	lastReloadAttemptTime time.Time
+	currentCertMutext     sync.RWMutex
+	currentCert           atomic.Pointer[tls.Certificate]
+}
+
+// module-wide log.
+var log zerolog.Logger
+
+// New creates a new cert manager service.
+func New(ctx context.Context, params ...Parameter) (*Service, error) {
+	parameters, err := parseAndCheckParameters(params...)
+	if err != nil {
+		return nil, errors.Wrap(err, "problem with parameters")
+	}
+
+	// Set logging.
+	log = zerologger.With().Str("service", "certmanager").Str("impl", "standard").Logger()
+	if parameters.logLevel != log.GetLevel() {
+		log = log.Level(parameters.logLevel)
+	}
+
+	// Load the certificates immediately.
+	certPEMBlock, err := parameters.majordomo.Fetch(ctx, parameters.certPEMURI)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to obtain server certificate")
+	}
+	certKeyBlock, err := parameters.majordomo.Fetch(ctx, parameters.certKeyURI)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to obtain server key")
+	}
+
+	// Initialise the certificate pair.
+	serverCert, err := tls.X509KeyPair(certPEMBlock, certKeyBlock)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to load certificate pair")
+	}
+	if len(serverCert.Certificate) == 0 {
+		return nil, errors.New("certificate file does not contain a certificate")
+	}
+	cert := serverCert.Leaf
```

Comment:
When I tested this the application panicked here as Leaf was nil. For the test setup I needed to replace with this:

```suggestion
	cert, err := x509.ParseCertificate(serverCert.Certificate[0])
	if err != nil || cert == nil {
		return nil, errors.Wrap(err, "failed to parse server certificate")
	}
```

I'm not sure if this caters for your usecase too or if we need to handle both `.Leaf` being populated and it not being.